dask_jobqueue/job.py

```diff
@@ -0,0 +1,480 @@
+            cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, **kwargs
+        )
+
+        out, err = proc.communicate()
+        out, err = out.decode(), err.decode()
+
+        if proc.returncode != 0:
+            raise RuntimeError(
+                "Command exited with non-zero exit code.\n"
+                "Exit code: {}\n"
+                "Command:\n{}\n"
+                "stdout:\n{}\n"
+                "stderr:\n{}\n".format(proc.returncode, cmd_str, out, err)
+            )
+        return out
+
+
+class JobQueueCluster(SpecCluster):
+    __doc__ = """ Deploy Dask on a Job queuing system
+
+    This is a superclass, and is rarely used directly.  It is more common to
+    use an object like SGECluster, SLURMCluster, PBSCluster, LSFCluster, or
+    others.
+
+    However, it can be used directly if you have a custom ``Job`` type.
+    This class relies heavily on being passed a ``Job`` type that is able to
+    launch one Job on a job queueing system.
+
+    Parameters
+    ----------
+    Job : Job
+        A class that can be awaited to ask for a single Job
+    {cluster_parameters}
+    """.format(
+        cluster_parameters=cluster_parameters
+    )
+
+    job_cls = None
+
+    def __init__(
+        self,
+        n_workers=0,
+        job_cls: Job = None,
+        # Cluster keywords
+        loop=None,
+        security=None,
+        silence_logs="error",
+        name=None,
+        asynchronous=False,
+        # Scheduler keywords
+        interface=None,
+        host=None,
+        protocol="tcp://",
+        dashboard_address=":8787",
+        config_name=None,
+        # Job keywords
+        **kwargs
+    ):
+        self.status = "created"
+        if job_cls is not None:
+            self.job_cls = job_cls
+
+        if self.job_cls is None:
+            raise ValueError(
+                "You must provide a Job type like PBSJob, SLURMJob, "
+                "or SGEJob with the job_cls= argument."
+            )
+
+        if config_name:
+            if interface is None:
+                interface = dask.config.get("jobqueue.%s.interface" % config_name)
+
+        scheduler = {
+            "cls": Scheduler,  # Use local scheduler for now
+            "options": {
+                "protocol": protocol,
+                "interface": interface,
+                "host": host,
+                "dashboard_address": dashboard_address,
+                "security": security,
+            },
+        }
+        if config_name:
+            kwargs["config_name"] = config_name
+        kwargs["interface"] = interface
+        kwargs["protocol"] = protocol
+        kwargs["security"] = security
+        self._kwargs = kwargs
+        worker = {"cls": self.job_cls, "options": kwargs}
+        if "processes" in kwargs and kwargs["processes"] > 1:
+            worker["group"] = ["-" + str(i) for i in range(kwargs["processes"])]
+
+        self.example_job  # trigger property to ensure that the job is valid
```

Comment:
I have not completely understood the magic behind `.example_job` yet ... I understand its main role is to implement `cluster.job_script` and you need to have the scheduler address for `.cluster_script` to be correct. Is there more to understand?

Reply:
Not really.  Users today are trained to look, at `job_script`.  We want to make it easy to print this out.  All of the logic for this is actually within the Job class now, so we want an easy way to generate a Job object.  

In this particular line we want to create a Job with all of our keywords because that is where all of the error handling and informative error messages are.  However, also at this point, we don't yet know the scheduler address, so we have to place in a dummy address.

Reply:
Is this ok with you @lesteve ?  Should I hit resolve here?  

Is there some comment I could add to the code that would make this more clear?

Reply:
Some suggestions:
* rename the variable to `._dummy_job` (better name more than welcome!) with a comment above saying that it is mostly to implement `JobQueueCluster.job_script`
* add a comment in `._dummy_job` about the `AttributeError` case (more or less what you replied in https://github.com/dask/dask-jobqueue/pull/307#discussion_r324386131). I still haven't fully understood your comment but mentioning that this is related to async may help